proof/refine/ARM/Ipc_R.thy

```diff
@@ -5199,12 +5265,180 @@ lemma cancelIPC_notin_epQueue:
+  "replyRemoveTCB param_a \<lbrace>\<lambda>s. P (scTCBs_of s)\<rbrace>"
+  unfolding replyRemoveTCB_def
+  apply (wpsimp wp: setSchedContext_scTCBs_of gts_wp')
+  apply (erule back_subst[where P=P], rule ext, clarsimp)
+  by (clarsimp simp: opt_map_def obj_at'_real_def ko_wp_at'_def)
+
+lemma replyRemoveTCB_tcb_scs_of'[wp]:
+  "replyRemoveTCB param_a \<lbrace>\<lambda>s. P (tcb_scs_of' s)\<rbrace>"
+  unfolding replyRemoveTCB_def
+  by (wpsimp wp: setSchedContext_scTCBs_of gts_wp')
+
+lemma replyRemoveTCB_tcbs_scs_sym_refs[wp]:
+  "replyRemoveTCB param_a \<lbrace>tcbs_scs_sym_refs\<rbrace>"
+  by (rule hoare_weaken_pre, wps, wpsimp, simp)
+
+crunches cancelIPC
+  for tcbs_scs_sym_refs[wp]: tcbs_scs_sym_refs
+  (wp: crunch_wps threadSet_tcb_scs_of'_inv
+ ignore: setSchedContext threadSet)
+
+lemma cleanReply_replies_scs_sym_refs :
+  "\<lbrace>\<lambda>s. sqheap_refs_inv (scReplies_of s) (\<lambda>a. if a = rptr then None else replySCs_of s a)\<rbrace>
+  cleanReply rptr
+  \<lbrace>\<lambda>_. replies_scs_sym_refs\<rbrace>" (is "\<lbrace>?pre\<rbrace> _ \<lbrace>_\<rbrace>")
+  unfolding cleanReply_def
+  apply (clarsimp simp: bind_assoc updateReply_def)
+  apply (wpsimp wp: hoare_drop_imp | wps)+
+  done
+
+lemma replyRemoveTCB_replies_scs_sym_refs [wp]:
+  "\<lbrace>replies_scs_sym_refs and (\<lambda>s. sym_refs (list_refs_of_replies' s))\<rbrace>
+  replyRemoveTCB t
+  \<lbrace>\<lambda>_. replies_scs_sym_refs\<rbrace>" (is "\<lbrace>?pre\<rbrace> _ \<lbrace>_\<rbrace>")
+  supply if_split [split del]
+  unfolding replyRemoveTCB_def
+  apply (clarsimp simp: bind_assoc updateReply_def)
+  apply wpsimp
+         apply (wpsimp wp: cleanReply_replies_scs_sym_refs)
+        apply wp
+         apply wps
+         apply wpsimp
+        apply (rule_tac Q="\<lambda>_ s. (replySCs_of s) (the (replyPrev reply)) = None \<and>
+                 sqheap_refs_inv (scReplies_of s)
+                  (\<lambda>a. if a = rptr then None else replySCs_of s a)"
+       in hoare_strengthen_post[rotated], clarsimp)
+        apply wpsimp
+       apply (rule_tac Q="\<lambda>_ s. (replyPrev reply \<noteq> None \<longrightarrow>
+             reply_at' (the (replyPrev reply)) s \<longrightarrow> (replySCs_of s) (the (replyPrev reply)) = None) \<and>
+                sqheap_refs_inv (scReplies_of s)
+                 (\<lambda>a. if a = rptr then None else replySCs_of s a)"
+      in hoare_strengthen_post[rotated])
+        apply (clarsimp split: if_splits simp: obj_at'_def)
+       apply (wp hoare_vcg_if_lift2 hoare_vcg_imp_lift' hoare_vcg_all_lift)
+         apply wps
+         apply (wpsimp wp: setSchedContext_scReplies_of)
+        apply wpsimp
+       apply wp
+        apply wps
+        apply wpsimp
+       apply wpsimp
+      apply wpsimp
+     apply (rule_tac Q="\<lambda>replya s. replies_scs_sym_refs s \<and> sym_refs (list_refs_of_replies' s)"
+    in hoare_strengthen_post[rotated])
+      apply (clarsimp split: if_splits)
+    apply (safe)
```

Comment:
Is `safe` generally preferred? I think I remember @Xaphiosis preferring `inro conjI impI` or something similar (if there is no better option)